### PR TITLE
Volume/shape based conflict detection for focal search

### PR DESCRIPTION
**Pull request by akmaralAW**
It would be helpful to know how would you solve it @quimortiz, this is what I am doing to check two motion primitives (one is a node being checked for expansion, second one is from the solution of other robots).

### CODE REVIEW

File: include/dynoplan/tdbastar/tdbastar_epsilon.hpp [hunk 1]
```diff
@@ -48,12 +48,20 @@ typedef typename boost::heap::d_ary_heap<
     boost::heap::compare<compareFocalHeuristic>, boost::heap::mutable_<true>>
     focal_t;
 
+void from_solution_to_motions_to_result(
+    std::vector<std::pair<std::shared_ptr<AStarNode>, size_t>> &result,
+    const std::vector<Motion> &motions, std::shared_ptr<AStarNode> solution_n,
+    std::map<size_t, Motion *> &map_motions_out);
+
```
Comment (whoenig) [inline include/dynoplan/tdbastar/tdbastar_epsilon.hpp:55]: this function is not used here, nor downstream?

File: src/tdbastar/tdbastar_epsilon.cpp [hunk 2]
```diff
@@ -178,6 +264,9 @@ void tdbastar_epsilon(
     Out_info_tdb &out_info_tdb, size_t &robot_id, bool reverse_search,
     std::vector<dynobench::Trajectory> &expanded_trajs,
     std::vector<LowLevelPlan<dynobench::Trajectory>> &solution,
+    std::vector<std::vector<std::pair<std::shared_ptr<AStarNode>, size_t>>>
+        &results,
+    std::vector<std::map<size_t, Motion *>> &result_motions,
```
[the remark this answers is not in the document]
Comment (whoenig) [reply src/tdbastar/tdbastar_epsilon.cpp:269]: Also, these are added, but never filled/used.

**Comment by quimortiz** [comment]
can you first explain to me what is focal heuristic? When you check two primitives are they "static", in the sense that we erase the notion of time and just check if their traces collides?

**Comment by quimortiz** [comment]
in the current status i cannot review the code: 1) I don't know what it is supposed to do. 2) Can you add some comments, e.g., says what is each argument to your new functions. 3) Is the code working and running correctly?

File: src/tdbastar/tdbastar_epsilon.cpp [hunk 3]
```diff
@@ -517,9 +606,12 @@ void tdbastar_epsilon(
...
       int focalHeuristic =
-          best_node->focalHeuristic +
-          lowLevelfocalHeuristic(solution, tmp_node, robot_id, all_robots,
-                                 col_mng_robots, robot_objs);
+          best_node->focalHeuristic; // + lowLevelfocalHeuristicShape(results,
+                                     // lazy_traj, robot_id, all_robots);
+                                     // lowLevelfocalHeuristic(solution,
+                                     // tmp_node, robot_id, all_robots,
+                                     //                      col_mng_robots,
+                                     //                      robot_objs);
```
[the remark this answers is not in the document]
Comment (akmaralAW) [reply src/tdbastar/tdbastar_epsilon.cpp]: addressed

File: src/tdbastar/tdbastar_epsilon.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (akmaralAW) [reply src/tdbastar/tdbastar_epsilon.cpp]: addressed

**Comment by akmaralAW** [comment]
@quimortiz, comments are added to the script. I will merge it to ecbs branch for now. 
Answering to your questions: 
1) this heuristic function is used to count the number of conflicts for the discrete search. It checks each motion primitive, that is considered as applicable for expansion, for collision with the solution/result of neighbor robots. 
2) Comments are added 
3) Yes, it is working and runs. At the moment I call it inside tdbA*-epsilon for focal set maintenance.

**Pull request merged**